src/components/Button/Button.tsx

```diff
@@ -0,0 +1,154 @@
+import React from 'react';
+import {
+  ActivityIndicator,
+  StyleSheet,
+  Text,
+  TouchableOpacity,
+} from 'react-native';
+import {ButtonProps} from './ButtonProps.types';
+import {scale, verticalScale} from '../../utils';
+import {useThemeColor} from '../../hooks';
+import Icon from 'react-native-vector-icons/Ionicons';
+
+export const Button = (props: ButtonProps) => {
+  const {
+    children,
+    mode,
+    isLoading,
+    style,
+    textStyle,
+    disabled,
+    onPress,
+    icon,
+    iconRight,
+    iconStyle,
+    loaderColor,
+    ...restProps
+  } = props;
+
+  const {colors} = useThemeColor();
+
+  const getStyle = () => {
+    if (mode === 'default') {
+      return;
+    } else if (mode === 'outlined') {
+      return [
+        styles.outlied,
+        {
+          borderColor: colors.primary,
+        },
+      ];
+    } else if (mode === 'text') {
+      return styles.textType;
+    }
+  };
+
+  const getDisabledBtnStyle = () => {
+    if (disabled && mode === 'text') {
+      return null;
+    } else if (disabled && mode === 'default') {
+      return {backgroundColor: colors.disabled};
+    } else if (disabled && mode === 'outlined') {
+      return {borderColor: colors.disabled};
+    }
+  };
+
+  const getDisabledLabelStyle = () => {
+    if (mode === 'default') {
+      return;
+    } else {
+      return {color: 'grey'};
+    }
+  };
+
+  return (
+    <TouchableOpacity
+      style={[
+        [
+          styles.buttonContainer,
+          {
+            backgroundColor: colors.primary,
+          },
+        ],
+        getStyle(),
+        getDisabledBtnStyle(),
+        iconRight ? styles.rightIcon : null,
+        style,
+      ]}
+      disabled={disabled || isLoading}
+      onPress={onPress}
+      {...restProps}>
+      {isLoading ? (
+        <ActivityIndicator size="small" color={loaderColor} />
+      ) : (
+        <>
+          {icon ? (
+            <Icon
+              name={icon}
+              style={[
+                styles.icon,
+                {
+                  color: mode === 'default' ? colors.text2 : colors.text,
+                },
+                disabled ? getDisabledLabelStyle() : null,
+                iconStyle,
+              ]}
+            />
+          ) : null}
+          <Text
+            style={[
+              styles.text,
+              mode !== 'default' ? {color: colors.onPrimary} : null,
+              {
+                color: mode === 'default' ? colors.text2 : colors.text,
```

Comment:
These conditions for color and disabled are very similar for icon and text. can we reuse them through functions or memo?

Reply:
extracted to a function and also used `useMemo` hook to memoized the values.